sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -155,47 +162,69 @@ private static <T> Mono<Pair<List<Range<String>>,QueryInfo>> getPartitionKeyRang
             });
     }
 
-    private static <T> Mono<Pair<List<Range<String>>, QueryInfo>> getTargetRangesFromQueryPlan(
+    private static <T> Mono<PartitionKeyRangesAndQueryInfos> getTargetRangesFromQueryPlan(
         CosmosQueryRequestOptions cosmosQueryRequestOptions, DocumentCollection collection,
         DefaultDocumentQueryExecutionContext<T> queryExecutionContext,
         PartitionedQueryExecutionInfo partitionedQueryExecutionInfo, Instant planFetchStartTime,
         Instant planFetchEndTime) {
+
         QueryInfo queryInfo =
             partitionedQueryExecutionInfo.getQueryInfo();
-        queryInfo.setQueryPlanDiagnosticsContext(new QueryInfo.QueryPlanDiagnosticsContext(planFetchStartTime,
-            planFetchEndTime,
-            partitionedQueryExecutionInfo.getQueryPlanRequestTimeline()));
+        if (queryInfo != null) {
+            queryInfo.setQueryPlanDiagnosticsContext(new QueryInfo.QueryPlanDiagnosticsContext(planFetchStartTime,
+                planFetchEndTime,
+                partitionedQueryExecutionInfo.getQueryPlanRequestTimeline()));
+        }
         List<Range<String>> queryRanges =
             partitionedQueryExecutionInfo.getQueryRanges();
 
         if (isScopedToSinglePartition(cosmosQueryRequestOptions)) {
```

Comment:
we should also make sure the query plan is not cached if customer using any feedRange/pkValue filter - non blocking, just need to follow up